xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -1902,94 +1940,400 @@ func (s) TestRingHash_ContinuesConnectingWithoutPicksOneSubchannelAtATime(t *tes
+			"address_hash", newFirstAddr+"_0",
+		))
+		if _, err := client.EmptyCall(rpcCtx, &testpb.Empty{}, grpc.Peer(&remote)); err != nil {
+			t.Fatalf("rpc EmptyCall() failed: %v", err)
+		}
+		if remote.Addr.String() == initialFirstAddr {
+			continue
+		}
+		t.Fatalf("Request went to unexpected backend %q, want backend %q", remote.Addr, initialFirstAddr)
 	}
-	if holdNonExistent2Again.IsStarted() {
-		t.Errorf("Got connection attempt to backend 2 again, expected no connection attempt.")
+}
+
+// Tests that requests are sent to the next address within the same endpoint
+// after the first address becomes unreachable.
+func (s) TestRingHash_FallBackWithinEndpoint(t *testing.T) {
+	origDualstackEndpointsEnabled := envconfig.XDSDualstackEndpointsEnabled
+	defer func() {
+		envconfig.XDSDualstackEndpointsEnabled = origDualstackEndpointsEnabled
+	}()
+	envconfig.XDSDualstackEndpointsEnabled = true
+	backends := startTestServiceBackends(t, 4)
+	backendAddrs := backendAddrs(backends)
+
+	xdsServer, nodeID, xdsResolver := setupManagementServerAndResolver(t)
+
+	const clusterName = "cluster"
+	endpoints := endpointResourceForBackendsWithMultipleAddrs(t, clusterName, [][]string{{backendAddrs[0], backendAddrs[1]}, {backendAddrs[2], backendAddrs[3]}})
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		ServiceName: clusterName,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+	route := channelIDHashRoute("new_route", virtualHostName, clusterName)
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	if err := xdsServer.Update(ctx, xdsUpdateOpts(nodeID, endpoints, cluster, route, listener)); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	const numRPCs = 5
+	received := checkRPCSendOK(ctx, t, client, numRPCs)
+	if len(received) != 1 {
+		t.Errorf("Got RPCs routed to %v backends, want %v", len(received), 1)
+	}
+	var got int
+	var initialAddr string
+	for initialAddr, got = range received {
+	}
+	if got != numRPCs {
+		t.Errorf("Got %v RPCs routed to a backend, want %v", got, numRPCs)
+	}
+
+	var otherEndpointAddr string
+	var backendIdx int
+	switch initialAddr {
+	case backendAddrs[0]:
+		otherEndpointAddr = backendAddrs[1]
+		backendIdx = 0
+	case backendAddrs[1]:
+		otherEndpointAddr = backendAddrs[0]
+		backendIdx = 1
+	case backendAddrs[2]:
+		otherEndpointAddr = backendAddrs[3]
+		backendIdx = 2
+	case backendAddrs[3]:
+		otherEndpointAddr = backendAddrs[2]
+		backendIdx = 3
+	default:
+		t.Fatalf("Request sent to unknown backend: %q", initialAddr)
```

Comment:
I don't quite understand this. Shouldn't the first RPC always reach the first address in the endpoint. i.e `backendAddrs[0]`?